Make sure custom relations work out with new alias strictness.

dereuromark opened the pull request:
Refs https://github.com/cakephp/bake/issues/1012

So basically 5.1 and the more strict relation-handling ("Association alias `Countries` is already set") kinda broke the non documented feature of custom relationship baking.
Even though it was just baked wrongly with always the same Countries relation.

This PR adds the expected aliasing (back) ensuring there is no collision, and having the correct name in the process for most cases:
```
[
  'belongsTo' => [
    (int) 0 => [
      'alias' => 'Countries',
      'foreignKey' => 'country_id',
      'className' => 'Countries'
    ],
    (int) 1 => [
      'alias' => 'BillingCountries',
      'foreignKey' => 'billing_country_id',
      'className' => 'Countries'
    ],
    (int) 2 => [
      'alias' => 'ShippingCountries',
      'foreignKey' => 'shipping_country_id',
      'className' => 'Countries'
    ]
  ],
  'hasOne' => [],
  'hasMany' => [],
  'belongsToMany' => []
]
```
as example for

Countries
- id

Users
- id
- country_id
- billing_country_id
- shipping_country_id

dereuromark committed:
Make sure custom relations work out with new alias strictness.

dereuromark committed:
Make sure custom relations work out with new alias strictness.

markstory commented on src/Command/ModelCommand.php:
```diff
@@ -1546,4 +1549,41 @@ protected function bakeEnums(Table $model, array $data, Arguments $args, Console
+     * @param array<string, mixed> $association
+     * @return string
+     */
+    protected function alias(array $association): string
```
```suggestion
    protected function createAssociationAlias(array $association): string
```

`alias` is a pretty general term, I think a more specific name could be helpful for future us.

markstory commented on src/Command/ModelCommand.php:
```diff
@@ -1546,4 +1549,41 @@ protected function bakeEnums(Table $model, array $data, Arguments $args, Console
+            foreach ($associationsPerType as $k => $association) {
+                $alias = $association['alias'];
+                if (in_array($alias, $existing, true)) {
+                    $alias = $this->alias($association);
```
```suggestion
                    $alias = $this->createAssociationAlias($association);
```

dereuromark committed:
Make sure custom relations work out with new alias strictness.

dereuromark commented:
It actually works:
```
        $this->belongsTo('Countries', [
            'foreignKey' => 'country_id',
            'className' => 'Countries',
        ]);
        $this->belongsTo('BillingCountries', [
            'foreignKey' => 'billing_country_id',
            'className' => 'Countries',
        ]);
        $this->belongsTo('ShippingCountries', [
            'foreignKey' => 'shipping_country_id',
            'className' => 'Countries',
        ]);
```

Maybe someone wants to add a test case for it? E.g. @luizcmarin ?
You can also test it and confirm

dereuromark committed:
Make sure custom relations work out with new alias strictness.

dereuromark committed:
Make sure custom relations work out with new alias strictness.

dereuromark commented:
Turns out there is already an existing test case that so far was already invalid (exception would have happened), and now working:
https://github.com/cakephp/bake/pull/1015/commits/9631dd049a4e63e5274b8a52c757d01c7ea96057

ADmad commented on tests/TestCase/Command/ModelCommandTest.php:
```diff
@@ -602,9 +602,10 @@ public function testGetAssociationsConstraints()
             ],
             [
-                'alias' => 'Users',
+                'alias' => 'Receivers',
```
Would be nice if the aliases were always created based on the foreign key field, so `Senders` for `sender_id` too instead of `Users`.

dereuromark replied:
It seems to be that way - until table lookup.
The problem seems to be that

                    $found = $this->findTableReferencedBy($schema, $fieldName);
                    if ($found) {
                        $tmpModelName = Inflector::camelize($found);
                    }
                    
finds the users table, and as such then switches it back to Users

dereuromark committed:
Make sure custom relations work out with new alias strictness.

dereuromark commented:
@ADmad I found a way to make it work.

luizcmarin committed:
Update ModelCommand.php

correction due to the error shown below.


Creating file C:\marin\www\GitHub\lixo\tests\TestCase\Model\Table\AnnouncementsTableTest.php
Wrote `C:\marin\www\GitHub\lixo\tests\TestCase\Model\Table\AnnouncementsTableTest.php`
Done
One moment while associations are detected. 2024-11-26 10:24:32 warning: Undefined variable $className
Trace:
Bake\Command\ModelCommand->findBelongsTo() C:\marin\www\GitHub\trash\vendor\cakephp\bake\src\Command\ModelCommand.php, line 247
Bake\Command\ModelCommand->getAssociations() C:\marin\www\GitHub\trash\vendor\cakephp\bake\src\Command\ModelCommand.php, line 166
Bake\Command\ModelCommand->getTableContext() C:\marin\www\GitHub\trash\vendor\cakephp\bake\src\Command\ModelCommand.php, line 119
Bake\Command\ModelCommand->bake() C:\marin\www\GitHub\lixo\vendor\cakephp\bake\src\Command\ModelCommand.php, line 101
Bake\Command\ModelCommand->execute() C:\marin\www\GitHub\trash\vendor\cakephp\bake\src\Command\AllCommand.php, line 116
Bake\Command\AllCommand->execute() C:\marin\www\GitHub\trash\vendor\cakephp\cakephp\src\Console\BaseCommand.php, line 204
Cake\Console\BaseCommand->run() C:\marin\www\GitHub\trash\vendor\cakephp\cakephp\src\Console\CommandRunner.php, line 330
Cake\Console\CommandRunner->runCommand() C:\marin\www\GitHub\trash\vendor\cakephp\cakephp\src\Console\CommandRunner.php, line 168
Cake\Console\CommandRunner->run() C:\marin\www\GitHub\lixo\bin\cake.php, line 10
[main]

luizcmarin commented:
> Maybe someone wants to add a test case for it? E.g. @luizcmarin ? You can also test it and confirm

#1018

dereuromark committed:
Merge pull request #1018 from luizcmarin/patch-1

Update ModelCommand.php

dereuromark committed:
Make sure custom relations work out with new alias strictness.

dereuromark committed:
Add more tests.

luizcmarin commented:
I created a small test with some tables where errors occur. You may want to take note of this inconvenience.

[invoices.txt](https://github.com/user-attachments/files/17932993/invoices.txt)
![image](https://github.com/user-attachments/assets/1917b089-cde1-4a09-a33a-49086d233266)

dereuromark commented:
I wonder if thats directly related to the issue here (was working before and now not)
It rather looks like a different topic and probably should be checked out independently (e.g. after merge of this).

luizcmarin commented:
> I wonder if thats directly related to the issue here (was working before and now not) It rather looks like a different topic and probably should be checked out independently (e.g. after merge of this).

In fact, bake never finished generating the files due to this problem. This also happened with other tables. I isolated these attached tables for you to test, if you wish. But it happens with others too. Does this happen to you?

I'll send you the full script, in case you want to try it.

[kancats.txt](https://github.com/user-attachments/files/17933707/kancats.txt)

dereuromark commented:
You need to actually share the commands you use
because the SQL table itself does not provide the info to replicate

For me both "bake all Countries" and then "bake all Users" works just fine now with the upgrade here.
Same for "Proposals" and "ProposalItems" of this schema.

So whats the concrete command that breaks, and what have been generated before, or if manually adjusted, what does that file look like before running bake.

Again: I am fairly certain this is outside the scope of this specific issue here, and should not block the relation fix for the current bug in the system.

luizcmarin commented:
I do not rule out the possibility of errors here. Please, if everything is ok in your tests, continue. I will do new tests after updating.

The command I use is simply: " bin\cake bake all --everything". Nothing more than that.

dereuromark commented:
I can partically reproduce it
The issue you are having is probably due to some invalid setup in your DB schema

        $this->hasMany('Proposals', [
            'foreignKey' => 'proposal_id',
            'className' => 'Invoices',
        ]);
        
This wrongly is a proposal_id, but should be something else at least to make clear this is from invoices table referenced, and also must always be different from the actual proposal_id since Proposals is their own table.
So you are creating on purpose a setup that bake everything cannot solve for you.
You need to fix your table setup here, e.g. use `invoice_proposal_id` or alike to make it unique.

luizcmarin commented:
Thank you very much for the tip. This means more work for me and less for you :) :(...
I'll review my scheme. I think this solves everything.

dereuromark commented:
Well, https://github.com/cakephp/bake/pull/1019 will make it easier for you once merged :)

dereuromark commented:
All good? Can we squash merge?

ADmad commented on src/Command/ModelCommand.php:
```diff
@@ -1546,4 +1554,43 @@ protected function bakeEnums(Table $model, array $data, Arguments $args, Console
+     * @param array<string, array<string, mixed>> $associations
+     * @return array<string, array<string, mixed>>
+     */
+    protected function ensureAliasUniqueness(array $associations): array
```
```suggestion
    protected function ensureUniqueAlias(array $associations): array
```
This rolls a bit easier off the tongue :)

dereuromark replied:
But it does it for more than one, so we could debate `ensureUniqueAliases()` etc.

dereuromark replied:
I must say I kinda like the naming ensureAliasUniqueness() that @markstory proposed.

ADmad replied:
Okay, we can keep the same then.

dereuromark replied:
Oh createAssociationAlias() was from him, not this one.
Either way, if all are fine, we can keep it.

ADmad reviewed (commented):
Looks good apart from the minor nitpick of function name.
